vmhost/contexts/asyncLocal.go

```diff
@@ -326,45 +302,46 @@ func (context *asyncContext) createCallbackInput(
 			OriginalTxHash:       runtime.GetOriginalTxHash(),
 			PrevTxHash:           runtime.GetPrevTxHash(),
 			ReturnCallAfterError: returnWithError,
+			ESDTTransfers:        lastTransferInfo.lastESDTTransfers,
 		},
-		RecipientAddr: context.address,
+		RecipientAddr: caller,
 		Function:      callbackFunction,
 	}
 	context.SetAsyncArgumentsForCallback(contractCallInput, asyncCall, gasAccumulated)
 
-	if isESDTOnCallBack {
-		context.updateContractInputForESDTOnCallback(contractCallInput, esdtFunction, esdtArgs, vmOutput, asyncCall, gasAccumulated)
-	}
-
 	return contractCallInput, nil
 }
 
-func (context *asyncContext) updateContractInputForESDTOnCallback(
-	contractCallInput *vmcommon.ContractCallInput,
-	esdtFunction string,
-	esdtArgs [][]byte,
-	vmOutput *vmcommon.VMOutput,
-	asyncCall *vmhost.AsyncCall,
-	gasAccumulated uint64) {
-
-	oldArgLen := len(contractCallInput.Arguments)
-	oldFunction := contractCallInput.Function
+func (context *asyncContext) extractLastTransferWithoutData(caller []byte, vmOutput *vmcommon.VMOutput) lastTransferInfo {
+	callValue := big.NewInt(0)
 
-	contractCallInput.Function = esdtFunction
-	contractCallInput.Arguments = make([][]byte, 0, oldArgLen)
-	contractCallInput.Arguments = append(contractCallInput.Arguments, esdtArgs...)
-	contractCallInput.Arguments = append(contractCallInput.Arguments, []byte(oldFunction))
-	contractCallInput.Arguments = append(contractCallInput.Arguments, ReturnCodeToBytes(vmOutput.ReturnCode))
+	callBackReceiver := context.host.Runtime().GetContextAddress()
+	outAcc, ok := vmOutput.OutputAccounts[string(callBackReceiver)]
+	if !ok || len(outAcc.OutputTransfers) == 0 {
```

Comment:
add || len(vmOutput.ReturnData) > 0

Reply:
emptyLastTransferInfo = lastTransferInfo{
			callValue:         big.NewInt(0),
			lastESDTTransfers: nil,
		}

And in case of early exits return it.

Reply:
done